### PR TITLE
Optionally output iteration loss as info log

### PR DESCRIPTION
**Pull request by torsjonas**
I am running implicit als inside a Docker container as a custom AWS SageMaker algorithm. As it is now, the loss during fit iterations is not picked up as a metric properly by the SageMaker log parser when the loss is outputted using tqmd (which has a funky output format). This PR adds an optional parameter (default False) to the als fit function to also output each loss as a regular info log. This would make it possible for AWS SageMaker to pick up loss during iterations as a metric from the regular info logs.

What do you think about the idea to optionally output plain info logs?

Ideally I would like to choose either tqmd or plain info logs, but to my understanding if show_progress is False the whole function call will be skipped. So outputting individual info logs would only work when also showing progress with tqmd.

Thanks for a great module!

### CODE REVIEW

**Comment by benfred** [comment]
Rather than add a parameter - can we just always log the training loss if calculate_training_loss=True and show_progress=False?

I'd also like to see the loss passed to the fit_callback function if calculate_training_loss=True so that you can inject arbitrary logic here to log the training loss rather than having to parse the loss from a log message. I used the 'fit_callback' functionality to graph training loss per iteration as part of this post https://www.benfrederickson.com/fast-implicit-matrix-factorization/ - the code to do this is here:

https://github.com/benfred/implicit/blob/471194c8774ec4cd2c9fae294d113762fc505962/benchmarks/benchmark_als.py#L29-L51

this involved calculating the loss myself though, and I think it would be nicer to have it passed to the fit_callback as a parameter instead.

**Comment by torsjonas** [comment]
Makes sense to pass it to the callback. And, it turns out I was misinformed that tqdm would short circuit the function call if progress=False, so I will update the PR with this and see how it turns out.

**Comment by torsjonas** [comment]
In the latest commit I always log iteration loss when calculate_training_loss=True and show_progress=False. Maybe more controversially I also always pass loss as a new parameter to the fit_callback, even if calculate_training_loss=False in which case the loss will be None. Figured that was easier than to having to deal with different fit_callback function signatures depending on the value of calculate_training_loss.

Downside with calling fit_callback suddenly with a new positional argument is that it would be a breaking change, not sure how you feel about that? I updated the benchmark script to use the new loss argument instead of calculating it externally. But others would have to update their callback function signature if they update the implicit library. An alternative that would not be a breaking change would be to add another callback that specifically also takes the loss as an argument, something like "fit_callback_loss". What do you think?

**Review by benfred — APPROVED** [review]

**Comment by benfred** [comment]
thanks! this looks good
